Edit/New forms

ilya-burinskiy opened the pull request.

Commit:
Edit/New forms

Commit:
Implementing actions and views for questions

Commit:
App footer and helpers

Commit:
Test resources

Commit:
Errors shared template

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -1,25 +1,35 @@
 
   def index
-    render json: @test.questions
+    @questions = @test.questions
```
по заданию Список вопросов теста должен выводиться на той же странице, что и сам тест
Зачем вообще этот экшен нужен?

VidgarVii replied:
этот экшен надо не фиксить, а удалять

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -0,0 +1,22 @@
+  private
+
+  def all_tests
+    @tests = Test.all
```
это строка должна быть в методе index. Код должен быть явный. Колбеки ухудшают читаемость кода

VidgarVii commented on app/helpers/application_helper.rb:
```diff
@@ -1,2 +1,9 @@
 module ApplicationHelper
+  def year
+    Time.zone.now.year
```
хорошо. А есть еще рельсовая реализация получения времени с учётом тайм зоны если она настроена. Time.current

VidgarVii commented on app/helpers/application_helper.rb:
```diff
@@ -1,2 +1,9 @@
+  end
+
+  def github_url(author, repo)
+    'https://github.com/' + author + '/' + repo
```
интерполяция строк на много эффективней и читается лучше

VidgarVii commented on app/views/layouts/_footer.html.erb:
```diff
@@ -0,0 +1,9 @@
+</p>
+<p>
+  Эта система - одно из заданий онлайн-школы <%= link_to "ThinkNetica", "https://thinknetica.com", rel: "nofollow", target: "_blank" %>
+</p>
```
фрагменты не должны лежить в папке лаяут. лучше положить в shared/

VidgarVii commented on app/views/questions/_form.html.erb:
```diff
@@ -0,0 +1,10 @@
+  <%= render 'shared/errors', resource: @question %>
+  <p>
+    <%= form.label :question %><br>
+    <%= form.text_field :body %>
```
так как поля дублируются, при изменении и создании объекта, их выносят в отдельный фрагмент

VidgarVii commented on app/views/questions/index.html.erb:
```diff
@@ -0,0 +1,9 @@
+
+<ul>
+  <%= render @questions %>
+</ul>
```
этот файл не нужен

VidgarVii commented on app/views/tests/_test.html.erb:
```diff
@@ -0,0 +1,5 @@
+<tr>
+  <td><%= test.title %></td>
+  <td><%= test.level %></td>
+  <td><%= test.questions.count %></td>
```
test.questions.size

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  def show
+    redirect_to test_questions_path(@test)
```
это оверхед. Не должно быть ни каких подобных редиректов. Тут следует возвращать тест и его вопросы

Commit:
Moved footer to shared dir

Commit:
Question controller and views fix

- Deleted index action and view
- Added show action and view
- Updated back link in new and edit forms

Commit:
Test controller and views fix

- Deleted all_tests callback
- Added show view
- Using test.question.count in test partial

Commit:
Question controller fix

- Updated redirect path in update and create actions

Commit:
Form fields partial

Commit:
Application helper fix

- Using Time.current.year
- Using interpolation for creating github url

VidgarVii commented on app/views/tests/show.html.erb:
```diff
@@ -0,0 +1,11 @@
+  <% @test.questions.each do |question| %>
+    <%= render 'questions/question', question: question %>
+  <% end %>
+</ul>
```
во всех ерб нет пустых строк в конце файла

VidgarVii commented on app/views/tests/index.html.erb:
```diff
@@ -0,0 +1,10 @@
+    <th>Questions</th>
+  </tr>
+  <%= render @tests %>
+</table>
```
семантика таблиц говорит, что должна быть шапка и тело таблицы
```
table
  thead
  rbody
```

VidgarVii commented on app/helpers/questions_helper.rb:
```diff
@@ -1,2 +1,8 @@
+
+  def question_header(test_title, mode)
+    MODS[mode] + ' ' + test_title + ' ' + 'Question'
+  end
```
про интерполяцию уже отписал. Конкатенация вообще плохой способ клеить строки

VidgarVii commented on app/views/tests/show.html.erb:
```diff
@@ -0,0 +1,11 @@
+</p>
+
+<ul>
+  <% @test.questions.each do |question| %>
```
плохая привычка строить запросы из представления. Тут надо работать только с теми данными, которые приходят из контроллера

VidgarVii commented on app/views/tests/show.html.erb:
```diff
@@ -0,0 +1,11 @@
+
+<ul>
+  <% @test.questions.each do |question| %>
+    <%= render 'questions/question', question: question %>
```
```
render 'questions/question', collection: questions
или
render @questions
```

Commit:
Query for test questions  in controller

Commit:
Using interpolation

Commit:
Terminating newlines

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -41,7 +46,7 @@ def find_test
 
   def find_question
-    @question = Question.find(params[:id])
+    @question = @test.questions.find(params[:id])
```
А если бы воспользовался бы shallow в роутах, то запросы были бы чище

http://rusrails.ru/rails-routing#melkoe-vlozhenie

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -0,0 +1,18 @@
+
+  def show
+    @questions = @test.questions
+  end
```
мог бы и колбек убрать. Он всё ровно оверхедом висит.

VidgarVii commented on app/views/questions/_form.html.erb:
```diff
@@ -0,0 +1,7 @@
+  <%= render 'shared/errors', resource: @question %>
+  <p><%= render 'form_fields', form: form %></p>
+
+  <%= form.hidden_field :test_id, value: params[:test_id] %>
```
это лишнее поле. test_id у тебя задается в model form_with

VidgarVii commented on app/views/tests/index.html.erb:
```diff
@@ -0,0 +1,13 @@
+    </tr>
+  </tbody>
+  <%= render @tests %>
+</table>
```
промахнулся. теперь у тебя шапка пустая.

VidgarVii replied:
не исправил

VidgarVii replied:
```
<table>
  <thead>
    <tr>
      <th>Title</th>
      <th>Level</th>
      <th>Questions</th>
    </tr>
  </thead>
  <tbody>
    <%= render @tests %>
  </tbody>
</table>
```

Commit:
Deleted test_id hidden field in question form

Commit:
Shallow true option in question routes

Commit:
Updated forms for question

Commit:
Updated routes in question partial and in show action

Commit:
Index view for question

Commit:
Updated questions controller

Commit:
Render tests in table body tag

Commit:
Find test callback also finds all tests

Commit:
Test controller fix

- Added find_tests and find_questions callbacks

VidgarVii commented on config/routes.rb:
```diff
@@ -1,6 +1,7 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
```
root так и не появился

VidgarVii commented on app/views/tests/_test.html.erb:
```diff
@@ -0,0 +1,5 @@
+  <td><%= test.title %></td>
+  <td><%= test.level %></td>
+  <td><%= test.questions.size %></td>
+</tr>
```
**В представлениях должны быть ссылки для навигации между страницами приложения**

должна быть возможность перехода м/у страницами. Т.е. надо добавить ссылку для перехода на страницу самого теста

VidgarVii commented on app/views/questions/edit.html.erb:
```diff
@@ -0,0 +1,8 @@
+
+  <p><%= form.submit %></p>
+<% end %>
+<%= link_to 'Back', test_path(@test) %>
```
лучше заюзать `link_to 'Back', :back`
а то не всегда угадаешь от куда пользователь пришел. Внутренних ссылок на одну страницу может быть несколько

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -0,0 +1,27 @@
+  before_action :find_test, only: %i[show]
+  before_action :find_tests, only: %i[index]
+  before_action :find_questions, only: %i[show]
+
```
колбек программирование убивает читаемость кода

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -1,39 +1,53 @@
+  def update
+    @test = @question.test
+    if @question.update(question_params)
+      redirect_to test_path(@test)
```
обычно после обновления какого-то объекта, возвращают на страницу самого объекта

Commit:
Deleted question index view and action

Commit:
Removed unnecessary callbacks in test controller

Commit:
Root route

Commit:
Test title has link to test

Commit:
Using link_to :back

Commit:
Test show view fix

- Back link leads to tests_url

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -1,39 +1,49 @@
   def destroy
+    @test = @question.test
     @question.destroy!
+    redirect_to test_path(@test)
```
переменная @test не используется в представлении. Да и самого представления нет. лучше сразу test_path(@question.test)

Commit:
Fixes
